frame/balances/src/lib.rs

```diff
@@ -944,6 +945,51 @@ impl<T: Config<I>, I: 'static> fungible::Unbalanced<T::AccountId> for Pallet<T,
 	}
 }
 
+impl<T: Config<I>, I: 'static> fungible::InspectReserve<T::AccountId> for Pallet<T, I> {
+	fn reserved_balance(who: &T::AccountId) -> T::Balance {
+		Self::account(who).reserved
+	}
+	fn total_balance(who: &T::AccountId) -> T::Balance {
+		Self::account(who).total()
+	}
+	fn can_reserve(who: &T::AccountId, amount: T::Balance) -> bool {
+		let a = Self::account(who);
+		let min_balance = T::ExistentialDeposit::get().max(a.frozen(Reasons::All));
+		if a.reserved.checked_add(&amount).is_none() { return false }
+		// We require it to be min_balance + amount to ensure that the full reserved funds may be
+		// slashed without compromising locked funds or destroying the account.
+		let required_free = match min_balance.checked_add(&amount) {
+			Some(x) => x,
+			None => return false,
+		};
+		a.free >= required_free
+	}
+}
+impl<T: Config<I>, I: 'static> fungible::MutateReserve<T::AccountId> for Pallet<T, I> {
+	fn reserve(who: &T::AccountId, amount: Self::Balance) -> DispatchResult {
+		if amount.is_zero() { return Ok(()) }
+		ensure!(Self::can_reserve(who, amount), Error::<T, I>::InsufficientBalance);
+		Self::mutate_account(who, |a| {
+			a.free -= amount;
+			a.reserved += amount;
+		})?;
+		Ok(())
+	}
+	fn unreserve(who: &T::AccountId, amount: Self::Balance) -> Result<T::Balance, DispatchError> {
+		if amount.is_zero() { return Ok(amount) }
+		// Done on a best-effort basis.
+		let actual = Self::mutate_account(who, |a| {
+			let new_free = a.free.saturating_add(amount.min(a.reserved));
+			let actual = new_free - a.free;
+			// ^^^ Guaranteed to be <= amount and <= a.reserved
+			a.free = new_free;
+			a.reserved = a.reserved.saturating_sub(actual.clone());
```

Comment:
is clone needed? this would be a surprise to me

Reply:
Only to make my editor not mark it red.